Bucket: expose Available() and Capacity()

hongchaodeng opened the pull request:
Hi,

The current token bucket implementation is great. Nonetheless we need some underlying data to implement more powerful functionalities.

For example, if we want to implement a function `CanAccept` which only checks available tokens without taking it; implement a function that exports rate limiter's metrics to better understand current flow and bottleneck on our services. Both functions will benefit if we can get the underlying data like `avail` and `capacity`, including the already public `Rate()`.

Let me know if there are questions. Highly appreciate it if anyone can help review this issue. Thanks!

axw commented on ratelimit.go:
```diff
@@ -171,6 +171,16 @@ func (tb *Bucket) takeAvailable(now time.Time, count int64) int64 {
 
+// Available returns the number of available tokens
+func (tb *Bucket) Available() int64 {
+	return tb.avail
```
tb.mu should be held while accessing tb.avail

axw commented:
Seems fine to me, with one small fix; deferring to @rogpeppe, the author, though.

xiang90 commented on ratelimit.go:
```diff
@@ -171,6 +171,18 @@ func (tb *Bucket) takeAvailable(now time.Time, count int64) int64 {
 }
 
+// Available returns the number of available tokens
+func (tb *Bucket) Available() int64 {
```
I guess we'd better also doc that `It will be negative when there are consumers waiting for tokens.`

rogpeppe commented on ratelimit.go:
```diff
@@ -171,6 +171,19 @@ func (tb *Bucket) takeAvailable(now time.Time, count int64) int64 {
+func (tb *Bucket) Available() int64 {
+	tb.mu.Lock()
+	defer tb.mu.Unlock()
+	return tb.avail
```
This is not right. The number of available tokens changes as time progresses
but the returned value from this function does not.

You need to adjust avail before returning the value. Something like
this:
	
	// Available returns the number of available tokens.
	// It will be negative when there are consumers waiting for tokens.
	// Note that this method is inherently racy - the returned
	// count may not be current a moment later.
	func (tb *Bucket) Available(now time.Time) int64 {
		return tb.available(time.Now())
	}
	
	// available is the internal version of available - it takes the current time as
	// an argument to enable easy testing.
	func (tb *Bucket) available(now time.Time) int64 {
		tb.mu.Lock()
		defer tb.mu.Unlock()
		tb.adjust(now)
		return tb.avail
	}

Also, please write a test (note the internal version above, as with the
other methods, written that way so it's easy to write a deterministic test).

rogpeppe commented on ratelimit.go:
```diff
@@ -171,6 +171,19 @@ func (tb *Bucket) takeAvailable(now time.Time, count int64) int64 {
+	return tb.avail
+}
+
+// Capacity returns the capacity of this rate limiter.
```
// Capacity returns the capacity that the bucket was created with.

rogpeppe commented:
Thanks for this.
This is good in principle, but could do with a little work to make it correct and tested.

Please rebase to a single commit before re-proposing, thanks.

hongchaodeng commented:
Thanks @rogpeppe for your reply. I realized my mistake and fixed it.
I have also added a test for the internal available().

I squash all I want to do in single commit. However, I add another commit to update `gocheck` and `math.Abs` -- the second one made the test not working.

Let me know if there are questions. Thanks again!

hongchaodeng commented:
@rogpeppe 
Changed naming to `TestAvailable()`

hongchaodeng commented:
@rogpeppe 
Addressed all comments. PTAL.

rogpeppe commented on ratelimit_test.go:
```diff
@@ -125,6 +126,32 @@ var takeTests = []struct {
+
+	expectCountAfterTake  int64
+	expectCountAfterSleep int64
+}{
```
}{{

to save an indent level.

Also, please move the tests to just before TestAvailable.

rogpeppe commented on ratelimit_test.go:
```diff
@@ -125,6 +126,32 @@ var takeTests = []struct {
+		capacity:              2,
+		fillInterval:          time.Second,
+		take:                  1,
+		sleep:                 time.Second,
```
I'd like to see at least one test that tests that available returns the correct thing when not called on a fillInterval boundary.

rogpeppe commented:
Thanks very much for making these changes. LGTM modulo some comments above.

hongchaodeng commented on ratelimit_test.go:
```diff
@@ -125,6 +126,44 @@ var takeTests = []struct {
+	sleep:                 time.Second,
+	expectCountAfterTake:  1,
+	expectCountAfterSleep: 2,
+}, { // shouldn't fill before interval
```
@rogpeppe 
Add two more tests to check boundaries. PTAL.

rogpeppe commented on ratelimit_test.go:
```diff
@@ -125,6 +126,44 @@ var takeTests = []struct {
+
+	expectCountAfterTake  int64
+	expectCountAfterSleep int64
+}{{ // should fill
```
let's make this comment into an "about" field for consistency with
other table-driven tests, and log it. (it means that when a test
does fail, we can quickly find out which one has failed)

rogpeppe commented:
LGTM with one trivial suggestion.

hongchaodeng commented on ratelimit_test.go:
```diff
@@ -125,6 +126,49 @@ var takeTests = []struct {
+	expectCountAfterTake  int64
+	expectCountAfterSleep int64
+}{{
+	about:                 "should fill tokens after interval",
```
@rogpeppe 
Add `about` in table and test failure output.

liggitt commented on ratelimit.go:
```diff
@@ -171,6 +171,28 @@ func (tb *Bucket) takeAvailable(now time.Time, count int64) int64 {
 
+// Available returns the number of available tokens.
+// It will be negative when there are consumers waiting for tokens.
+// Note that this method is inherently racy - the returned count
```
rather than "inherently racy" (since the available number is actually computed in a thread-safe way), maybe clarify that the returned number of available tokens does not guarantee calls like Take/TakeAvailable will succeed, and that this is primarily for introspection/metrics/reporting/debugging?

rogpeppe replied:
That's a good idea, thanks.
@hongchaodeng How about this?

// Available returns the number of available tokens.
// It will be negative when there are consumers waiting for tokens.
// Note that if this returns greater than zero,
// it does not guarantee that calls that take tokens
// from the buffer will succeed, as the number of available
// tokens could have changed in the meantime.
// This method is intended primarily for metrics reporting
// and debugging.

rogpeppe replied:
Actually, I'll land this now and make that change in another PR.

rogpeppe replied:
Done.
